libnavigation-core/src/main/java/com/mapbox/navigation/core/telemetry/MapboxNavigationTelemetry.kt

```diff
@@ -0,0 +1,296 @@
+
+    /**
+     * The delegate lambda that distaches either a pre or after initialization logic
+     */
+    private var postEventDelegate = postEventBeforeInit
+
+    /**
+     * One-time initializer. Called in responce to initialize() and then replaced with a no-op lambda to prevent multiple initialize() calls
+     */
+    private val primaryInitializer: (Context, String, MapboxNavigation, LocationEngine) -> Unit = { context, token, mapboxNavigation, locationEngine ->
+        this.context = context
+        mapboxToken = token
+        this.locationEngine = locationEngine
+        val options = MapboxNavigationOptions.Builder().build()
+        validateAccessToken(mapboxToken)
+        initializer = postInitialize // prevent primaryInitializer() from being called more than once.
+        postEventDelegate = postEventAfterInit // now that the object has been initialized we can post events
+
+        registerForNotification(mapboxNavigation)
+        monitorChannels()
+        val mapboxTelemetry = MapboxTelemetry(context, token, obtainUserAgent(options))
+        mapboxTelemetry.enable()
+
+        /**
+         * Register a callback to receive location events. At most [MAX_LOCATION_VALUES] are stored
+         */
+        locationEngine.requestLocationUpdates(locationEngineRequest, locationCallback, null)
+    }
+    private var initializer = primaryInitializer
+    private var postInitialize: (Context, String, MapboxNavigation, LocationEngine) -> Unit = { _, _, _, _ -> }
+    fun initialize(
+        context: Context,
+        mapboxToken: String,
+        mapboxNavigation: MapboxNavigation,
+        locationEngine: LocationEngine
+    ) {
+        initializer(context, mapboxToken, mapboxNavigation, locationEngine)
+    }
+
+    /**
+     * This method is used to post all types of telemetry events to the back-end server.
+     * The [event] parameter represents one of several Telemetry events avialable
+     */
+    override fun postTelemetryEvent(event: TelemetryEventInterface) {
+        postEventDelegate(event)
+    }
+
+    private fun postEvent(event: TelemetryEventInterface) {
+        jobControl.scope.launch {
+            withContext(ThreadController.IODispatcher) { populateTelemetryEventWrapper(event) }?.let { telemetryEvent ->
+                updateEventsQueue(telemetryEvent)
+            }
+        }
+    }
+
+    private fun updateEventsQueue(telemetryEvent: TelemetryEventInterface) {
+        telemetryEventsQueue.add(telemetryEvent)
+        if (telemetryEventsQueue.size >= MAX_TELEMTRY_EVENTS) {
+            telemetryEventsQueue.forEach { event ->
+                mapboxTelemetry.push(event as Event)
+            }
+            telemetryEventsQueue.clear()
+        }
+        telemetryEventsQueue.add(telemetryEvent)
+    }
+
+    private fun registerForNotification(mapboxNavigation: MapboxNavigation) {
+        mapboxNavigation.registerRouteProgressObserver(routeProgressListener)
```

Comment:
See comment above re: listening for updates from `MapboxNavigation`.

Reply:
👍